packages/zoe/src/contracts/newSwap/swap.js

```diff
@@ -0,0 +1,303 @@
+      want: {
+        Out: { brand: brandOut },
+      },
+    } = seat.getProposal();
+    const { brand: brandIn } = amountIn;
+
+    // we could be swapping (1) central to secondary, (2) secondary to
+    // central, or (3) secondary to secondary
+
+    if (isCentral(brandIn) && isSecondary(brandOut)) {
+      const brandOutPool = getPool(brandOut);
+      const pool = getPool(brandOut);
+      const {
+        amountOut,
+        amountIn: reducedAmountIn,
+        protocolFee,
+      } = getPriceGivenAvailableInput(amountIn, brandOut);
+
+      const seatStage = seat.stage({
+        In: amountMath.subtract(amountIn, reducedAmountIn),
+        Out: amountOut,
+      });
+      const poolStage = pool.getPoolSeat().stage({
+        Central: amountMath.add(
+          brandOutPool.getCentralAmount(),
+          amountMath.subtract(reducedAmountIn, protocolFee),
+        ),
+        Secondary: amountMath.subtract(
+          brandOutPool.getSecondaryAmount(),
+          amountOut,
+        ),
+      });
+      zcf.reallocate(poolStage, seatStage, stageProtocolSeatFee(protocolFee));
+
+      seat.exit();
+      getPool(brandOut).updateState();
+      return `Swap successfully completed.`;
+    } else if (isSecondary(brandIn) && isCentral(brandOut)) {
+      const brandInPool = getPool(brandIn);
+      // this branch is very similar to the above, with only pool and the pool's
+      // gains and losses changing. Sharing code makes it less readable.
+      const pool = getPool(brandIn);
+      // The protocolFee comes out of amountOut, which is provided by the pool
+      const {
+        amountOut,
+        amountIn: reducedAmountIn,
+        protocolFee,
+      } = getPriceGivenAvailableInput(amountIn, brandOut);
+
+      const seatStage = seat.stage({
+        In: amountMath.subtract(amountIn, reducedAmountIn),
+        Out: amountOut,
+      });
+      const poolStage = pool.getPoolSeat().stage({
+        Central: amountMath.subtract(
+          brandInPool.getCentralAmount(),
+          amountMath.add(amountOut, protocolFee),
+        ),
+        Secondary: amountMath.add(
+          brandInPool.getSecondaryAmount(),
+          reducedAmountIn,
+        ),
+      });
+      zcf.reallocate(poolStage, seatStage, stageProtocolSeatFee(protocolFee));
+
+      seat.exit();
+      getPool(brandIn).updateState();
+      return `Swap successfully completed.`;
+    } else if (isSecondary(brandIn) && isSecondary(brandOut)) {
+      // The protocol fee is extracted from the proceeds of the Input pool
+      // before depositing the remainder to the Output pool
+      const brandInPool = getPool(brandIn);
+      const brandOutPool = getPool(brandOut);
+
+      const {
+        amountIn: reducedAmountIn,
+        amountOut,
+        protocolFee,
+        // @ts-ignore two pool version of getPrice returns central
+        centralAmount,
+      } = getPriceGivenAvailableInput(amountIn, brandOut);
+
+      const seatStage = seat.stage(
```

Comment:
This too could be cleanly shared.

Reply:
possibly not worth it if we need to convert this to virtual pools

Reply:
done